### DISCUSSION TITLE
Custom compilation issue

### DISCUSSION THREAD

**Discussion by Wolll** [arendst/Tasmota#12723] [Q&A] [answered]
Hello
For some time now, Tasmota version 9.5.0.2 has been doing development arendst / Tasmota when compiling.
when I try to compile Tasmota32 webcam with my user_config_override I get the following error message

maybe an idea what was changed, before everything went smoothly 

Gruß Woolli

Building in release mode
*** use provided user_config_override.h as planned ***
*** use provided platformio_override.ini as planned ***
*** use provided platformio_tasmota_cenv.ini as planned ***
Use manually specified: COM4
Compiling .pio\build\tasmota32-webcam\src\homekit.c.o
xtensa-esp32-elf-gcc: error: unrecognized command line option '-mfix-esp32-psram-cache-strategy=memw'
Compiling .pio\build\tasmota32-webcam\src\tasmota.ino.cpp.o
Retrieved `.pio\build\tasmota32-webcam\partitions.bin' from cache
xtensa-esp32-elf-g++: error: unrecognized command line option '-mfix-esp32-psram-cache-strategy=memw'
*** [.pio\build\tasmota32-webcam\src\homekit.c.o] Error 1
*** [.pio\build\tasmota32-webcam\src\tasmota.ino.cpp.o] Error 1
Compiling .pio\build\tasmota32-webcam\liba0b\FS\FS.cpp.o
xtensa-esp32-elf-g++: error: unrecognized command line option '-mfix-esp32-psram-cache-strategy=memw'
*** [.pio\build\tasmota32-webcam\liba0b\FS\FS.cpp.o] Error 1

**Comment by sfromis** [comment, marked as answer]
A common fix for build issues to clean the PlatformIO setup, by removing the "hidden" folder `.platformio`

**Comment by Wolll** [comment]
I downloaded the version on 7/10/21 that still worked.
The versions after that no longer worked.

**Reply by ascillato** [reply]
That is because your platform.io files (the Arduino core and others) are old.
You need to delete the .platformio folder in order to force Platform.io download latest versions.

**Reply by ghost** [reply]
Not an solution, no other Projekt have this compiler Problems. I don't want to delete this folder everytime the error occures!
It take so many time to download all the files from the Internet, that is not an solution!
If an Update is needed, Plattform.IO should Update itself, what it everytime does,  not with this Brute Force method. 

Retrieved `.pio\build\tasmota32-core2\partitions.bin' from cache
xtensa-esp32-elf-g++: error: CreateProcess: No such file or directory
Compiling .pio\build\tasmota32-core2\libf27\lvgl\core\lv_obj.c.o
*** [.pio\build\tasmota32-core2\src\tasmota.ino.cpp.o] Error 1

or this, after deleting the cache

Compiling .pio\build\tasmota32-core2\src\tasmota.ino.cpp.o
xtensa-esp32-elf-g++: error: CreateProcess: No such file or directory
Compiling .pio\build\tasmota32-core2\libf27\lvgl\core\lv_obj.c.o
*** [.pio\build\tasmota32-core2\src\tasmota.ino.cpp.o] Error 1


This are errors that very often appear with tasmota and now we have 1-2024. Nothing is happen to resolve  this issue on Plattform IO or Tasmota.
All the "helpfull" hints on the Web are usless or only an short time solution, delete this or delete there is not an solution!
I delete the cache I delete the .plattformIO I delete also only the toolchain-xtensa-esp32, sometimes it helps sometimes not! And the problem is that this error appears again in weeks or months which should not happen!

**Reply by Jason2866** [reply]
Not an Tasmota issue. Can't reproduce either. If you can provide a reproducable way to force this behaviour open an issue in repo https://github.com/platformio/platform-espressif32/issues

**Reply by ghost** [reply]
May be may be not, but if there is no real solution for it, why every time it will recommended to delete the whole  "PlatformIO setup, by removing the "hidden" folder .platformio" ??? 

In my case I only delete all the folders in my "packages" folder which have an additionally "@ and Version Informations" in his name and thats it. I can Compile again without any error. The problem is that I can't reproduce this there because I have not create an screenshot of the folder before I deleted it. I do not know where this folders come from. I can only wait until the errors came back.  

In my Packages folder there are folders like tool-mkfatfs or  toolchain-xtensa-esp32s3 and some of them were there two times but with an extension named  "@ versionnumbers" 
like eg. tools-mklittlefs **@ 8.0.1** and so on. I don't know where they come from. 

I will have to wait until this errors appear again and then I will take the screenshots and as many informations I can get.

**Reply by Jason2866** [reply]
Screenshots what is installed doesnt help anything. To track down the problem it needs a workflow.
If I do this and after that this I encounter the issue. If you can provide this, and it is reproducible on other installations there is a chance to find the reason for.

The folders come from when different versions of toolchains or tools or platform packages are installed. This is normal.

Tasmota has its own Arduino framework, since some functions are not available in official one.
So if you switch between official and Tasmota framework you will find already two in your package folder. Since mklittlefs is not working anymore since core 1.x. I updated the tool and provide this one in Tasmota platform. This is officially fully supported from Platformio, so that's why I wrote open a ticket in Platformio github when issues occur. If you compile Tasmota for the esp32-solo1 you will find another framework in your package folder ;-)

Tasmota is not "just" a piece of software. All used backend Tools are modified and optimized for Tasmota. And there is support for MCUs (C2 and C6) which are not supported from official Platformio Platform.

**Comment by Wolll** [comment]
platform.io updates everything OK
thanks

**Comment by Wolll** [comment]
Hello
Tasmota version from 07/20/21 is running.
Tasmota version from 07/23/21 brings an error message again when compiling 


Compiling .pio\build\tasmota32-webcam\lib4e8\WiFi\WiFiGeneric.cpp.o
Compiling .pio\build\tasmota32-webcam\lib4e8\WiFi\WiFiMulti.cpp.o
D:/Eigene Dateien/Arduino-Programme/Sonoff-Tasmota/Sonoff-Tasmota-arends-23.7.21/tasmota/xdrv_10_scripter.ino: In function 'char* isvar(char*, uint8_t*, T_INDEX*, float*, char*, GVARS*)':
D:/Eigene Dateien/Arduino-Programme/Sonoff-Tasmota/Sonoff-Tasmota-arends-23.7.21/tasmota/xdrv_10_scripter.ino:2231:13: error: 'LITTLEFS' was not declared in this scope
             LITTLEFS.format();
             ^
Compiling .pio\build\tasmota32-webcam\lib4e8\WiFi\WiFiSTA.cpp.o
Compiling .pio\build\tasmota32-webcam\lib4e8\WiFi\WiFiScan.cpp.o
Compiling .pio\build\tasmota32-webcam\lib4e8\WiFi\WiFiServer.cpp.o
*** [.pio\build\tasmota32-webcam\src\tasmota.ino.cpp.o] Error 1

**Reply by Jason2866** [reply]
It is a regression from the change LITTLEFS a few days ago to LittleFS for the ESP32.
Is fixed with commit https://github.com/arendst/Tasmota/commit/923be119a2d9479696e883b23bcc130fe1b3778b

**Comment by Wolll** [comment]
thanks

**Comment by Wolll** [comment]
Hello
Still got an error message when compiling 


Retrieved `.pio\build\tasmota32-webcam\lib96c\default\be_modtab.c.o' from cache
Retrieved `.pio\build\tasmota32-webcam\lib96c\default\be_path_tasmota_lib.c.o' from cache
D:/Eigene Dateien/Arduino-Programme/Sonoff-Tasmota/Sonoff-Tasmota-arends-23.7.21/tasmota/tasmota.ino:29:25: fatal error: t_bearssl.h: No such file or directory

*******************************************************************
* Looking for t_bearssl.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:t_bearssl.h"
* Web  > https://platformio.org/lib/search?query=header:t_bearssl.h
*
*******************************************************************

   #include <t_bearssl.h>                    // We need to include before "tasmota_globals.h" to take precedence over the BearSSL version in Arduino
                        ^
compilation terminated.
Retrieved `.pio\build\tasmota32-webcam\lib96c\default\be_port.cpp.o' from cache
*** [.pio\build\tasmota32-webcam\src\tasmota.ino.cpp.o] Error 1
Retrieved `.pio\build\tasmota32-webcam\lib96c\default\be_tasmotalib.c.o' from cache
Retrieved `.pio\build\tasmota32-webcam\lib96c\default\be_webserver_lib.c.o' from cache
============================================================================================ [FAILED] Took 11.74 seconds ============================================================================================
Environment       Status    Duration
----------------  --------  ------------
tasmota32-webcam  FAILED    00:00:11.742

**Comment by sfromis** [comment]
With standard build, adding new features like TLS is not just so supported for tasmota32-webcam, only for tasmota32.

**Comment by Wolll** [comment]
when I go to the platformio_tasmota_env32.ini
make a change
from 

[env:tasmota32-webcam]
extends                 = env:tasmota32_base
board                   = esp32-cam
build_flags             = ${env:tasmota32_base.build_flags} -DFIRMWARE_WEBCAM
lib_extra_dirs          = lib/libesp32, lib/lib_basic

on
[env:tasmota32-webcam]
extends                 = env:tasmota32_base
board                   = esp32-cam
build_flags             = ${env:tasmota32_base.build_flags} -DFIRMWARE_WEBCAM
lib_extra_dirs          = lib/libesp32, lib/lib_basic, lib/lib_i2c, lib/lib_rf, lib/lib_div, lib/lib_ssl, lib/lib_display 

then I have no error message.
Can't you include that directly.

**Reply by sfromis** [reply]
Doing such additions is not very likely to be included in standard build setup. In this case due to webcam already being hard work for the ESP32, and adding TLS is also pretty hard work for such a small MCU. Basically such a combo is not really recommended, but you are free to try what you can make work. Just don't expect much sympathy in case of it not working well.

**Reply by Jason2866** [reply]
> when I go to the platformio_tasmota_env32.ini
> make a change
> from
> 
> [env:tasmota32-webcam]
> extends = env:tasmota32_base
> board = esp32-cam
> build_flags = ${env:tasmota32_base.build_flags} -DFIRMWARE_WEBCAM
> lib_extra_dirs = lib/libesp32, lib/lib_basic
> 
> on
> [env:tasmota32-webcam]
> extends = env:tasmota32_base
> board = esp32-cam
> build_flags = ${env:tasmota32_base.build_flags} -DFIRMWARE_WEBCAM
> lib_extra_dirs = lib/libesp32, lib/lib_basic, lib/lib_i2c, lib/lib_rf, lib/lib_div, lib/lib_ssl, lib/lib_display
> 
> then I have no error message.
> Can't you include that directly.

No, it will not. We choose very carefully the features (for every build) which are included.
Since webcam is already a heavy task for the ESP32, TLS is definitly a no go to include.
